Add support for resources and test resources in IDEA module

fanick1 opened the pull request:
Issue: #1738

Signed-off-by: Frantisek Veverka <fanick1@users.noreply.github.com>

### Context
currently the idea plugin  is missing support for specifying resource directories in the idea module

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/.github/CONTRIBUTING.md)
- [x] Make sure that all commmits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation
- [ ] Recognize contributor in release notes

fanick1 commented:
Hello, @oehme. I see that you have previously merged some changes regarding the IDEA plugin, so I figured you might be the right person to perhaps give me feedback on this PR.

bmuschko commented:
@donat Is this something you could look into?

fanick1 commented:
Hello @oehme, @bmuschko, @donat is there anything I could personally do to help move this forward?

oehme commented:
@fanick1 I'm sorry, we're pretty tightly scheduled at the moment, but I'll schedule this PR for Gradle 4.7

donat commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/model/idea/IdeaContentRoot.java:
```diff
@@ -58,6 +58,29 @@
+     *
+     * @since 4.5
+     */
+    DomainObjectSet<? extends IdeaSourceDirectory> getGeneratedResourceDirectories();
```
Why do we need the `getGeneratedResourceDirectories()`  and `getGeneratedTestResourceDirectories ()` methods? Clients can just test the `isGenerated` attribute for each element.

fanick1 replied:
These methods are provided to keep the interface consistent - methods   `getSourceDirectories()`, `getTestDirectories()` also have their counterparts `getGeneratedSourceDirectories()`, getGeneratedTestDirectories().

donat commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/model/idea/IdeaContentRoot.java:
```diff
@@ -58,6 +58,29 @@
+    /**
+     * The set of resource directories.
+     */
+    DomainObjectSet<? extends IdeaSourceDirectory> getResourceDirectories();
```
Missing @Incubating

donat replied:
Missing `@since 4.6`

donat commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/model/idea/IdeaContentRoot.java:
```diff
@@ -58,6 +58,29 @@
+    /**
+     * The set of resource directories.
+     */
+    DomainObjectSet<? extends IdeaSourceDirectory> getResourceDirectories();
```
The [IDEA documentation](https://www.jetbrains.com/help/idea/working-with-modules.html) explicitly states that The resources and test resources directory is only available for Java projects. I think we should explicitly specify this use case in the javadoc.

fanick1 replied:
added a note to the javadoc

donat commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/model/idea/IdeaContentRoot.java:
```diff
@@ -58,6 +58,29 @@
+    /**
+     * The set of test resource directories.
+     */
+    DomainObjectSet<? extends IdeaSourceDirectory> getTestResourceDirectories();
```
Same comments as for `getResourceDirectories`.

fanick1 commented on subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/m5/ToolingApiIdeaModelCrossVersionSpec.groovy:
```diff
@@ -114,7 +115,8 @@ idea.module.testOutputDir = file('someTestDir')
         module.compilerOutput.testOutputDir == file('someTestDir')
     }
-
+    @ToolingApiVersion("<4.4")
```
By 4.6 you mean 4.7, correct? So all the version labels in  `@ToolingApiVersion` `@TargetGradleVersion` and `@since` should reflect the 4.7 version, right?

fanick1 replied:
rebased & squashed

donat replied:
Yes, I meant 4.7. Sorry for the confusion.

fanick1 commented on subprojects/docs/src/docs/dsl/org.gradle.plugins.ide.idea.model.IdeaModule.xml:
```diff
@@ -17,7 +17,7 @@
                 <td>sourceDirs</td>
                 <td><literal>[]</literal></td>
-                <td>source dirs from <literal>project.sourceSets.main.allSource</literal></td>
+                <td>source dirs from <literal>project.sourceSets.main.allJava</literal></td>
```
good catch, I'll fix this

donat replied:
👍

fanick1 commented on subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r45/ToolingApiIdeaModelCrossVersionSpec.groovy:
```diff
@@ -0,0 +1,63 @@
+
+    @ToolingApiVersion(">=4.5")
+    @TargetGradleVersion(">=4.5")
+    def "provides source dir information"() {
```
do you mean to have these three use cases covered as part of this cross version test ? The case with the multi-project build is covered by one of the integration tests, so I have added tests just for the first two ones

donat replied:
Ok, you are right. I think there's still one use-case missing: when the Tooling API calls the new methods on `IdeaContentRoot` with an older Gradle distribution. Like [this one](https://github.com/gradle/gradle/blob/a0307889bed8a9ce58e1faf005a86ad86e1fb99e/subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r210/ToolingApiEclipseModelCrossVersionSpec.groovy#L33-L42).

donat commented:
@fanick1 I must have missed the notification about your changes, sorry about that. My next task will be to do the review.

fanick1 commented:
Ok, I have fixed most of the missing `@Incubating` annotations and `@since` in JavaDocs, but I don't know what to do about class `org.gradle.plugins.ide.idea.model.Module` and its method `configure` to be specific. I have added two more input arguments to this method which the check reports as broken binary compatibility.  
Will you bless  this breaking change or should I find a way around this? 
I could possibly keep the original method and add separate one with the new arguments and call the old one from there, but I would rather not to.

oehme commented:
That method is an accidental leak, it's not meant to be public API, so you can accept the change.

donat commented on subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r47/ToolingApiIdeaModelCrossVersionSpec.groovy:
```diff
@@ -0,0 +1,203 @@
+class ToolingApiIdeaModelCrossVersionSpec extends ToolingApiSpecification {
+
+    @ToolingApiVersion(">=4.7")
+    @TargetGradleVersion(">=4.7")
```
It's not a blocking issue, but you can move these annotations to the class level. This way you can eliminate the duplication and you can still override the annotations if you specify different values for a specific method.

donat commented on subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/Module.java:
```diff
@@ -181,15 +221,20 @@ protected String getDefaultResourceName() {
     }
 
     protected Object configure(Path contentPath,
```
Declare the binary compatibility breakage in the `subprojects/distributions/src/changes/accepted-public-api-changes.json` file. The content should be something like this (I copied it from the compatibility check report):
```
{
    "type": "org.gradle.plugins.ide.idea.model.Module",
    "member": "Method org.gradle.plugins.ide.idea.model.Module.configure(org.gradle.plugins.ide.idea.model.Path,java.util.Set,java.util.Set,java.util.Set,java.util.Set,java.util.Set,java.util.Set,java.lang.Boolean,org.gradle.plugins.ide.idea.model.Path,org.gradle.plugins.ide.idea.model.Path,java.util.Set,java.lang.String,java.lang.String)",
    "acceptation": "This method is not meant to be part of the public API"
},
{
    "type": "org.gradle.plugins.ide.idea.model.Module",
    "member": "Method org.gradle.plugins.ide.idea.model.Module.configure(org.gradle.plugins.ide.idea.model.Path,java.util.Set,java.util.Set,java.util.Set,java.util.Set,java.lang.Boolean,org.gradle.plugins.ide.idea.model.Path,org.gradle.plugins.ide.idea.model.Path,java.util.Set,java.lang.String,java.lang.String)",
    "acceptation": "This method is not meant to be part of the public API",
    "changes": [
        "Method has been removed"
    ]
}
```

donat commented on subprojects/ide/src/crossVersionTest/groovy/org/gradle/plugins/ide/tooling/r47/ToolingApiIdeaModelCrossVersionSpec.groovy:
```diff
@@ -0,0 +1,203 @@
+        root.testResourceDirectories.size() == 1
+        root.testResourceDirectories[0].directory == file('testResources')
+    }
+
```
Add test coverage to check that all new methods throw `UnsupportedOperationException` when older Gradle distributions are used.

```
@TargetGradleVersion("<4.7")
...
when:
root.getResourceDirectories()

then:
thrown UnsupportedOperationException
```

donat commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/model/idea/IdeaContentRoot.java:
```diff
@@ -58,6 +58,40 @@
+     * @since 4.7
+     */
+    @Incubating
+    DomainObjectSet<? extends IdeaSourceDirectory> getResourceDirectories();
```
For all new methods declare that they throw `UnsupportedOperationException` when invoked on an older Gradle distribution.

fanick1 commented:
I have made requested changes.
 - `UnsupportedOperationException` declarations and its coverage in the cross version test for new methods in the `IdeaContentRoot`
 - declare binary compatibility breakage in `Module`

fanick1 commented:
@donat, please check out my latest changes

donat commented:
The latest version looks good! I'll request an internal review, and we'll merge this PR

oehme commented on subprojects/tooling-api/src/main/java/org/gradle/tooling/model/idea/IdeaContentRoot.java:
```diff
@@ -58,6 +59,40 @@
+     * @since 4.7
+     */
+    @Incubating
+    DomainObjectSet<? extends IdeaSourceDirectory> getGeneratedTestResourceDirectories() throws UnsupportedMethodException;;
```
I don't know why we started this weird pattern of having convenience methods to get the generated dirs, but I don't think we should continue it. I'd probably deprecate the existing ones. The Tooling API should have as little surface as we can get away with.

fanick1 replied:
I have removed my two new getGenerated* methods and marked the remaining two as deprecated.

oehme commented on subprojects/docs/src/docs/dsl/org.gradle.plugins.ide.idea.model.IdeaModule.xml:
```diff
@@ -17,7 +17,7 @@
                 <td>sourceDirs</td>
                 <td><literal>[]</literal></td>
-                <td>source dirs from <literal>project.sourceSets.main.allSource</literal></td>
+                <td>source dirs from <literal>project.sourceSets.main.allJava</literal></td>
```
❌ Maybe I'm missing something, but this looks like a breaking change. IntelliJ currently calls `getSourceDirs` to determine all source directories and they don't yet know about the new `getResourceDirs` method. That means Gradle 4.7 projects would fail to import into IntelliJ.

donat replied:
@fanick1 I'm sorry that I missed this, but @oehme is correct. This PR should be modified as follows:
- `sourceDirs` and `testSourceDirs` should have the original content (`allSource`).
- The documentation for the getter methods should explicitly say, that `allSource` is returned only for compatibility reasons and `allJava` will be returned in Gradle 5.0.

fanick1 replied:
I'look into this today. Should I also cover this future change in a cross version test annotated with target api ≥ 5.0?

oehme replied:
That's a great idea, because it'll remind us to change it when 5.0 comes around.

fanick1 replied:
fixed

blindpirate commented:
Unfortunately I have to revert it for now because there's a failure on CI... no idea why the previous builds passed. Leave for investigation.

https://builds.gradle.org/viewLog.html?buildId=11892043&tab=buildResultsDiv&buildTypeId=Gradle_Check_QuickFeedbackCrossVersion_Java7_Oracle_Linux_ide&logTab=
